src/modules/mavlink/mavlink_messages.cpp

```diff
@@ -5310,6 +5311,179 @@ class MavlinkStreamRawRpm : public MavlinkStream
 	}
 };
 
+class MavlinkStreamESCInfo : public MavlinkStream
+{
+public:
+	const char *get_name() const override
+	{
+		return MavlinkStreamESCInfo::get_name_static();
+	}
+
+	static constexpr const char *get_name_static()
+	{
+		return "ESC_INFO";
+	}
+
+	static constexpr uint16_t get_id_static()
+	{
+		return MAVLINK_MSG_ID_ESC_INFO;
+	}
+
+	uint16_t get_id() override
+	{
+		return get_id_static();
+	}
+
+	static MavlinkStream *new_instance(Mavlink *mavlink)
+	{
+		return new MavlinkStreamESCInfo(mavlink);
+	}
+
+	unsigned get_size() override
+	{
+		static constexpr unsigned size_per_batch = MAVLINK_MSG_ID_ESC_INFO_LEN + MAVLINK_NUM_NON_PAYLOAD_BYTES;
+		return _esc_status_sub.advertised() ? size_per_batch * number_of_batches : 0;
+	}
+
+private:
+	uORB::Subscription _esc_status_sub{ORB_ID(esc_status)};
+
+	/* do not allow top copying this class */
+	MavlinkStreamESCInfo(MavlinkStreamESCInfo &) = delete;
+	MavlinkStreamESCInfo &operator = (const MavlinkStreamESCInfo &) = delete;
+
+protected:
+	uint8_t number_of_batches{0};
+
+	explicit MavlinkStreamESCInfo(Mavlink *mavlink) : MavlinkStream(mavlink)
+	{}
+
+	bool send(const hrt_abstime t) override
+	{
+
+		esc_status_s esc_status;
+
+		uint8_t batch_size = MAVLINK_MSG_ESC_INFO_FIELD_TEMPERATURE_LEN;
+
+		if (_esc_status_sub.update(&esc_status)) {
+
+			mavlink_esc_info_t msg = {};
+
+			msg.time_usec = esc_status.timestamp;
+			msg.counter = esc_status.counter;
+			msg.count = esc_status.esc_count;
+			msg.connection_type = esc_status.esc_connectiontype;
+			msg.info = esc_status.esc_online_flags;
+
+			number_of_batches = (esc_status.esc_count + (batch_size - 1)) / batch_size;
+
+			for (int batch_number = 0; batch_number < number_of_batches; batch_number++) {
+
+				msg.index = batch_number * batch_size;
+
+				for (int esc_index = 0; esc_index < batch_size ; esc_index++) {
+
+					msg.failure_flags[esc_index] = esc_status.esc[esc_index].failures;
+					msg.error_count[esc_index] = esc_status.esc[esc_index].esc_errorcount;
+					msg.temperature[esc_index] = esc_status.esc[esc_index].esc_temperature;
+				}
+
+				mavlink_msg_esc_info_send_struct(_mavlink->get_channel(), &msg);
```

Comment:
I find the batch size calculation a bit unclear. Wouldn't it be simpler to take the `esc_count` divided by the number of slots in `ESC_INFO` (4)?

Reply:
The calculation of batch size or number of batches? The batch size will always be 4, i.e. the number of messages in a batch is 4. I just took the length of one of the fields in case the message changes.

if you mean number of batches we can do that and take the ceiling of the division. so that if there are 5 ESCs then we would need two batches with 4 + (1+3) messages (in the second batch only the first esc has data, the final 3 are empty).

Reply:
Yes that's what I meant. At first glance it wasn't obvious why the batch_size is `MAVLINK_MSG_ESC_INFO_FIELD_TEMPERATURE_LEN`.

Reply:
So should I change the calculation? Or perhaps include a comment to explain it? I can't remember where I found the formula but it does give the correct result, i.e. 1-4 esc = 1 batch; 5-8 = 2 batches etc

Reply:
Comment to explain or simpler logic that's self-explanatory.

Reply:
Done. 
Thanks